push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/sender/GCMPushNotificationSender.java

```diff
@@ -159,19 +160,44 @@ private void cleanupInvalidRegistrationIDsForVariant(String variantID, Multicast
             if (errorCodeName != null) {
                 logger.info(String.format("Processing [%s] error code from GCM response, for registration ID: [%s]", errorCodeName, registrationIDs.get(i)));
             }
-
-            // is there any 'interesting' error code, which requires a clean up of the registration IDs
-            if (GCM_ERROR_CODES.contains(errorCodeName)) {
-
-                // Ok the result at INDEX 'i' represents a 'bad' registrationID
-
-                // Now use the INDEX of the _that_ result object, and look
-                // for the matching registrationID inside of the List that contains
-                // _all_ the used registration IDs and store it:
-               inactiveTokens.add(registrationIDs.get(i));
+            
+            //after sending, lets find tokens that are inactive from now on and need to be replaced with the new given canonical id.
+            //according to gcm documentation, google refreshes tokens after some time. So the previous tokens will become invalid.
+            //When you send a notification to a registration id which is expired, for the 1st time the message(notification) will be delivered
+            //but you will get a new registration id with the name canonical id. Which mean, the registration id you sent the message to has 
+            //been changed to this canonical id, so change it on your server side as well.
+            
+            //check if current index of result has canonical id
+            String canonicalRegId = result.getCanonicalRegistrationId();
+            if (canonicalRegId != null) {
+                // same device has more than one registration id: update it
+              
+                // find the installation and change its token to new canonical id
+                Installation installation = clientInstallationService.findInstallationForVariantByDeviceToken(variantID,registrationIDs.get(i));
```

Comment:
@secondsun sent a PR @sinarz's branch, where I add a check for that.

Is it really impossible? I wonder if it's possible that (due to errors) one device has multiple (outdated) registration ids